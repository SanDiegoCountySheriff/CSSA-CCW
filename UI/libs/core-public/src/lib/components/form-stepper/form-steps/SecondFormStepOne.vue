<template>
  <div>
    <v-form
      ref="form"
      v-model="state.valid"
    >
      <v-subheader class="sub-header, font-weight-bold">
        {{ $t(' Employment Status') }}
      </v-subheader>

      <v-row class="ml-5">
        <v-col cols="6">
          <v-select
            id="select"
            :items="employmentStatus"
            :label="$t(' Employment Status')"
            :rules="[v => !!v || $t(' Employment status is required')]"
            @change="
              v => {
                completeApplication.employment = v.toLowerCase();
              }
            "
          >
          </v-select>
        </v-col>
      </v-row>
      <v-divider />
      <div v-if="completeApplication.employment === 'employed'">
        <v-subheader class="sub-header font-weight-bold">
          {{ $t('Work Information') }}
        </v-subheader>
        <v-row>
          <v-col>
            <v-text-field
              :label="$t('Employer Name')"
              :rules="[v => !!v || $t('You must enter a employer name')]"
              v-model="completeApplication.workInformation.employerName"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-text-field>

            <v-text-field
              :label="$t('Employer Address Line 1')"
              :rules="[v => !!v || $t('You must enter a address')]"
              v-model="completeApplication.workInformation.employerAddressLine1"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-text-field>
            <v-text-field
              :label="$t('Employer Address Line 2')"
              v-model="completeApplication.workInformation.employerAddressLine2"
            >
            </v-text-field>
            <v-text-field
              :label="$t('Employer City')"
              :rules="[v => !!v || $t('You must enter a city')]"
              v-model="completeApplication.workInformation.employerCity"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-text-field>
          </v-col>
          <v-col>
            <v-autocomplete
              :items="states"
              :label="$t('Employer State')"
              :rules="[v => !!v || $t('You must enter a state')]"
              v-model="completeApplication.workInformation.employerCity"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-autocomplete>
            <v-text-field
              :label="$t('Employer Zip Code')"
              :rules="[v => !!v || $t('You must enter a Zip Code')]"
              v-model="completeApplication.workInformation.employerZip"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-text-field>
            <v-text-field
              :label="$t('Employer Phone number')"
              :rules="[v => !!v || $t('You must enter a phone number')]"
              v-model="completeApplication.workInformation.employerPhone"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-text-field>
            <v-autocomplete
              :items="countries"
              :label="$t('Employer Country')"
              :rules="[v => !!v || $t('You must enter a country')]"
              v-model="completeApplication.workInformation.employerCountry"
            >
              <template #prepend>
                <v-icon
                  x-small
                  color="error"
                >
                  mdi-asterisk
                </v-icon>
              </template>
            </v-autocomplete>
          </v-col>
        </v-row>
      </div>
    </v-form>
    <div class="weapon-components-container">
      <WeaponsTable :weapons="completeApplication.weapons" />
      <WeaponsDialog :save-weapon="getWeaponFromDialog" />
    </div>
    <v-divider />
    <FormButtonContainer
      :valid="state.valid"
      @submit="updateMutation.mutate"
      @save="saveMutation.mutate"
    />
  </div>
</template>

<script setup lang="ts">
import FormButtonContainer from '@core-public/components/containers/FormButtonContainer.vue';
import WeaponsDialog from '@core-public/components/dialogs/WeaponsDialog.vue';
import WeaponsTable from '@shared-ui/components/tables/WeaponsTable.vue';
import { employmentStatus, states } from '@shared-utils/lists/defaultConstants';
import { i18n } from '@shared-ui/plugins';
import { reactive } from 'vue';
import { useCompleteApplicationStore } from '@core-public/stores/completeApplication';
import { useMutation } from '@tanstack/vue-query';
import { useRouter } from 'vue-router/composables';

interface ISecondFormStepOneProps {
  handleNextSection: CallableFunction;
}
const completeApplicationStore = useCompleteApplicationStore();
const completeApplication =
  completeApplicationStore.completeApplication.application;

const props = defineProps<ISecondFormStepOneProps>();
const router = useRouter();

const state = reactive({
  valid: false,
});

const updateMutation = useMutation({
  mutationFn: () => {
    return completeApplicationStore.updateApplication('Step six complete');
  },
  onSuccess: () => {
    props.handleNextSection();
  },
  onError: error => {
    alert(error);
  },
});

const saveMutation = useMutation({
  mutationFn: () => {
    return completeApplicationStore.updateApplication('Save and quit');
  },
  onSuccess: () => {
    router.push('/');
  },
  onError: () => {
    alert(i18n.t('Save unsuccessful, please try again'));
  },
});

function getWeaponFromDialog(weapon) {
  completeApplication.weapons.push(weapon);
}
</script>

<style lang="scss" scoped>
.weapon-components-container {
  display: flex;
  flex-direction: column;
  width: 100%;
  justify-content: flex-start;
  align-items: flex-start;
}
</style>
