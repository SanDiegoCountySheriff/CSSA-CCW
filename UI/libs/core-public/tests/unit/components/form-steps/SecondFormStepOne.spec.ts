import SecondFormStepOne from '@core-public/../../../../../shared/ui/src/lib/components/form-stepper/form-steps/SecondFormStepOne.vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

const pinia = createTestingPinia();
const localVue = createLocalVue();
const tMock = {
  $t: text => text,
};

describe('SecondFormStepOne', () => {
  let vuetify;
  let wrapper;

  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = mount(SecondFormStepOne, {
      localVue,
      vuetify,
      pinia,
      mocks: tMock,
      propsData: {
        handleNextSection: () => null,
      },
      data() {
        return {
          valid: false,
        };
      },
    });
  });

  it('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should render the header', () => {
    expect(wrapper.text()).toContain('Employment Status');
  });

  it('Should display the menu on click', async () => {
    await wrapper.find('#select').trigger('click');
    expect(wrapper.text()).toContain('Unemployed');
  });
});
