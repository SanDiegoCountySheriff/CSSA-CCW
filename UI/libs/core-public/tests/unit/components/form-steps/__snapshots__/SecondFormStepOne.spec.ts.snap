// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondFormStepOne Should match the snapshot 1`] = `
<div>
  <form novalidate="novalidate" class="v-form">
    <div class="v-subheader sub-header, font-weight-bold theme--light">
      Employment Status
    </div>
    <div class="row ml-5">
      <div class="col-lg-6 col-12">
        <div class="v-input theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-5" class="v-input__slot">
              <div class="v-select__slot"><label for="select" class="v-label theme--light" style="left: 0px; position: absolute;"> Employment Status</label>
                <div class="v-select__selections"><input id="select" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
    <!---->
  </form>
  <div class="weapon-components-container">
    <div class="container container--fluid">
      <div class="v-data-table text-left my-3 theme--light">
        <div class="v-data-table__wrapper">
          <table>
            <thead>
              <tr>
                <th>
                  Make
                </th>
                <th>
                  Model
                </th>
                <th>
                  Caliper
                </th>
                <th>
                  Serial Number
                </th>
              </tr>
            </thead>
            <tbody></tbody>
          </table>
        </div>
      </div>
    </div>
    <div><button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="add-weapon-btn" role="button" aria-haspopup="true" aria-expanded="false"><span class="v-btn__content">
        Add Weapon
      </span></button>
      <div class="v-dialog__container">
        <!---->
      </div>
    </div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="form-btn-container">
    <div class="form-btn-inner">
      <div><button type="button" disabled="disabled" class="mt-3 v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"><span class="v-btn__content">
        Continue
      </span></button> <button type="button" class="mt-3 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default secondary mx-2"><span class="v-btn__content">
        Save and Exit
      </span></button></div>
      <div><button type="button" class="mt-3 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default warning mx-2"><span class="v-btn__content">
        Go back
      </span></button> <button type="button" class="mt-3 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default error mr-2"><span class="v-btn__content">
        Cancel
      </span></button></div>
    </div>
  </div>
  <div class="v-snack v-snack--bottom">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet v-sheet--outlined theme--light error--text" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          Section update unsuccessful please try again.
        </div>
        <div class="v-snack__action "></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
